types/propEq.d.ts

```diff
@@ -1,6 +1,32 @@
-export function propEq<T>(val: T): {
-  <K extends PropertyKey>(name: K): (obj: Record<K, T>) => boolean;
-  <K extends PropertyKey>(name: K, obj: Record<K, T>): boolean;
+import { Placeholder } from 'ramda';
+import { WidenLiterals } from '../util/tools';
+
+export function propEq(__: Placeholder): never;
+export function propEq<V>(val: V): {
+  <K extends PropertyKey>(name: K): K extends number
+    ? (array: Array<WidenLiterals<V>>) => boolean
+    : (obj: Record<K, WidenLiterals<V>>) => boolean;
+  <const U>(__: Placeholder, obj: U): (name: keyof U) => boolean;
+  <K extends keyof U, const U>(name: K, obj: U): boolean;
```

Comment:
I don't think `const U` in needed here. `U` should be find.

You need to constrain `U` and `K` here though. If `typeof U[K]` as well. Turns out you don't need `WidenLiterals` here at all, in fact, it would actually break if `U[K]` was `string | number` but you passed a string to `val: V`.

See here: https://tsplay.dev/m3KbAm